[Doc] Update documentation for `uv run` (#51233)
ray-project/ray: doc/source/ray-core/handling-dependencies.rst

First, it keeps dependencies synchronized between your driver and Ray workers.
Additionally, it provides full support for `pyproject.toml` including editable
packages. It also allows you to lock package versions using `uv lock`.
For more details, see the `UV scripts documentation <https://docs.astral.sh/uv/guides/scripts/>`_ as
well as `our blog post <https://www.anyscale.com/blog/uv-ray-pain-free-python-dependencies-in-clusters>`_.

.. note::

>   Since this is a new feature, you currently need to set a feature flag:

Reviewer: ```suggestion
  Because this is a new feature, you currently need to set a feature flag:
```

.. testcode::
  :skipif: True

  import emoji
  import ray

  @ray.remote
  def f():
      return emoji.emojize('Python is :thumbs_up:')

  # Execute 1000 copies of f across a cluster.
  print(ray.get([f.remote() for _ in range(1000)]))

> and run the driver script with `uv run test.py`. This will run 1000 copies of

Reviewer: ```suggestion
and run the driver script with `uv run test.py`. This runs 1000 copies of
```

the `f` function across a number of Python worker processes in a Ray cluster.

> The `emoji` dependency, in addition to being available for the main script, will

Reviewer: ```suggestion
The `emoji` dependency, in addition to being available for the main script, is
```

> also be available for all worker processes. Also, the source code in the current

Reviewer: ```suggestion
also available for all worker processes. Also, the source code in the current
```

> working directory will be available to all the workers.

Reviewer: ```suggestion
working directory is available to all the workers.
```

This workflow also supports editable packages, for example, you can use
`uv add --editable ./path/to/package` where `./path/to/package`
must be inside your current working directory so it's available to all
workers.

See `here <https://www.anyscale.com/blog/uv-ray-pain-free-python-dependencies-in-clusters#end-to-end-example-for-using-uv>`_
for an end-to-end example of how to use `uv run` to run a batch inference workload
with Ray Data.

**Using uv in a Ray Job:** With the same `pyproject.toml` and `test.py` files as above,
you can submit a Ray Job via

.. code-block:: sh

  ray job submit --working-dir . -- uv run test.py

> This will make sure both the driver and workers of the job run in the uv environment as specified by your `pyproject.toml`.

Reviewer: ```suggestion
This command makes sure both the driver and workers of the job run in the uv environment as specified by your `pyproject.toml`.
```

**Using uv with Ray Serve:** With an appropriate `pyproject.toml` and `app.py` files, you can
run a Ray Serve application with `uv run serve run app:main`.

**Best Practices and Tips:**

- Use `uv lock` to generate a lockfile and make sure all your dependencies are frozen, so things won't change in uncontrolled ways if a new version of a package gets released.

> - If you have a requirements.txt file, you can use `uv add -r requirement.txt` to add the dependencies to your `project.toml` and then use that with uv run.

Reviewer 2: should this be pyproject.toml?

Author: Yes, I fixed it, thanks a lot for catching that!

- If your `pyproject.toml` is in some subdirectory, you can use `uv run --project` to use it from there.

> - If you use uv run and want to reset the working directory to something that is not the current working directory, use the `--directory` flag. The Ray uv integration will make sure your `working_dir` will be set accordingly.

Reviewer: ```suggestion
- If you use uv run and want to reset the working directory to something that isn't the current working directory, use the `--directory` flag. The Ray uv integration makes sure your `working_dir` is set accordingly.
```

driver. Also, the `working_dir` runtime environment is used to propagate the working directory of the driver
(including the `pyproject.toml`) to the workers. This allows uv to set up the right dependencies and environment for the
workers to run in. There are some advanced use cases where you might want to use the `py_executable` mechanism directly in
your programs:

- *Applications with heterogeneous dependencies:* Ray supports using a different runtime environment for different
  tasks or actors. This is useful for deploying different inference engines, models, or microservices in different
  `Ray Serve deployments <https://docs.ray.io/en/latest/serve/production-guide/handling-dependencies.html#dependencies-per-deployment>`_
  and also for heterogeneous data pipelines in Ray Data. To implement this, you can specify a
  different `py_executable` for each of the runtime environments and use uv run with a different
  `--project` parameter for each. Instead, you can also use a different `working_dir` for each environment.

- *Customizing the command the worker runs in:* On the workers, you might want to customize uv with some special

>   arguments that are not used for the driver. Or, you might want to run processes using `poetry run`, a build system

Reviewer: ```suggestion
  arguments that aren't used for the driver. Or, you might want to run processes using `poetry run`, a build system
```
